src/storage.rs

```diff
@@ -0,0 +1,103 @@
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    pin::Pin,
+    sync::{Arc, Mutex},
+};
+
+use automerge as am;
+use futures::Future;
+
+use super::DocumentId;
+
+pub trait Storage: Clone + Debug {
+    type Error: std::error::Error + Send + 'static;
+    type GetFuture: Future<Output = Result<Option<am::Automerge>, Self::Error>>;
+    type ListFuture: Future<Output = Result<Vec<DocumentId>, Self::Error>>;
+    type AppendFuture: Future<Output = Result<(), Self::Error>>;
+    type CompactFuture: Future<Output = Result<(), Self::Error>>;
+
+    fn get(&self, id: DocumentId) -> Self::GetFuture;
+    fn list(&self) -> Self::ListFuture;
+    fn append(&self, id: DocumentId, changes: &[u8]) -> Self::AppendFuture;
+    fn compact(&self, id: DocumentId) -> Self::CompactFuture;
+}
+
+#[derive(Clone, Debug)]
+pub struct InMemory {
+    docs: Arc<Mutex<HashMap<DocumentId, am::Automerge>>>,
+}
+
+impl InMemory {
+    pub fn new() -> Self {
+        Self {
+            docs: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+impl Storage for InMemory {
+    type Error = std::convert::Infallible;
+    type GetFuture = futures::future::Ready<Result<Option<am::Automerge>, Self::Error>>;
+    type ListFuture = futures::future::Ready<Result<Vec<DocumentId>, Self::Error>>;
+    type AppendFuture = futures::future::Ready<Result<(), Self::Error>>;
+    type CompactFuture = futures::future::Ready<Result<(), Self::Error>>;
+
+    fn get(&self, id: DocumentId) -> Self::GetFuture {
+        let docs = self.docs.lock().unwrap();
+        futures::future::ready(Ok(docs.get(&id).cloned()))
+    }
+
+    fn list(&self) -> Self::ListFuture {
+        let docs = self.docs.lock().unwrap();
+        futures::future::ready(Ok(docs.keys().cloned().collect()))
+    }
+
+    fn append(&self, id: DocumentId, changes: &[u8]) -> Self::AppendFuture {
+        let mut docs = self.docs.lock().unwrap();
+        let doc = docs.entry(id).or_insert_with(|| am::Automerge::new());
+        doc.load_incremental(changes).unwrap();
+        futures::future::ready(Ok(()))
+    }
+
+    fn compact(&self, id: DocumentId) -> Self::CompactFuture {
+        // nothing to do as we're not actually storing things
+        futures::future::ready(Ok(()))
+    }
```

Comment:
Just to clarify, so if a document has been previously saved using `save_incremental`, then the saved data can be compacted as is? I thought one would have had to overwrite the data with that coming from a call to `save`.

Reply:
The implementation would have to load up all the incremental saves and then compact them using `am::Automerge::{load, save}` yeah. The reason I've defined the trait in these terms though is that it leaves concurrency control up to the storage implementation. E.g. if you're storing everything in a postgres database you can just do everything in a transaction. If we defined things in terms of `Vec<u8>` and did the compaction in `spanreed` then we would have to add our own kind of representation of a transaction over the top. Something like `Storage::begin_compaction` and `Storage::complete_compaction`.